perses/rjmc/geometry.py

```diff
@@ -2149,6 +2162,115 @@ def _calculate_growth_idx(self, particle_indices, growth_indices):
+            The system containing the relevant forces and particles
+        growth_indices : list of atom
+            The order in which the atom indices will be proposed
+        parameter_name : str
+            The name of the global context parameter
+        add_extra_torsions : bool, optional
+            Whether to add additional torsions to keep rings flat. Default true.
+        force_names : list of str
+            A list of the names of forces that will be included in this system
+        force_parameters : dict
+            Options for the forces (e.g., NonbondedMethod : 'CutffNonPeriodic')
+        verbose : bool, optional, default=False
+            If True, will print verbose output.
+
+        # We assume `reference_system` remains unmodified
+
+        """
+        self.sterics_cutoff_distance = 9.0 * units.angstroms # cutoff for sterics
+
+        self.verbose = verbose
+
+        # Get list of particle indices for new and old atoms.
+        self._new_particle_indices = [ atom.idx for atom in growth_indices ]
+        self._old_particle_indices = [idx for idx in range(reference_system.getNumParticles()) if idx not in self._new_particle_indices]
+        self._growth_indices = growth_indices
+
+        # Determine forces to keep
+        forces_to_keep = ['HarmonicBondForce', 'HarmonicAngleForce', 'PeriodicTorsionForce']
+        if use_sterics:
+            force_to_keep += ['NonbondedForce']
+
+        # Create reference system, removing forces we won't use
+        self._reference_system = copy.deepcopy(reference_system)
+        force_indices_to_remove = list()
+        for force_index in range(self._reference_system.getNumForces()):
+            force = self._reference_system.getForce(force_index)
+            force_name = force.__class__.__name__
+            if force_name not in forces_to_keep:
+                force_indices_to_remove.append(force_index)
+        for force_index in force_indices_to_remove[::-1]:
+            self._reference_system.removeForce(force_index)
+
+        # Create new system, copying forces we will keep.
+        self._growth_system = copy.deepcopy(self._reference_system)
+
+        # Zero all parameters
+        self.set_growth_parameter_index(0)
+
+        # TODO: Add extra ring-closing torsions, if requested.
+
+        # TODO: Precompute growth indices for force terms for speed
+
+    def set_growth_parameter_index(self, growth_index, context=None):
+        """
+        Set the growth parameter index
+        """
+        for (growth_force, reference_force) in zip(self._growth_system.getForces(), self._reference_system.getForces()):
+            force_name = growth_force.__class__.__name__
+            if (force_name == 'HarmonicBondForce'):
+                for bond in range(reference_force.getNumBonds()):
+                    parameters = reference_force.getBondParameters(bond)
+                    this_growth_index = self._calculate_growth_idx(parameters[:2], self._growth_indices)
```

Comment:
You can precalculate these growth indices for each of the bonds, angles, torsions, particles, and exceptions in order to further speed this up.

If you're always incrementing the growth index by 1, you can also save time by only updating the parameters that are changing.